### ISSUE TITLE
[no-unnecessary-type-assertion] Exhaustiveness check in TS requires `!`

### ISSUE THREAD

**Issue by Skalman** [typescript-eslint/typescript-eslint#1282]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```TS
type A = { kind: "abc" } | { kind: "def" };

function f(a: A) {
  switch (a.kind) {
    case "abc":
      return 1;
    case "def":
      return 2;
    default:
      // This `!` is necessary:
      throw new Error("Bad kind: " + a!?.kind);
  }
}
```

**Expected Result**
No error, since the `!` is necessary.

**Actual Result**

> "This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion"

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.9.0` |
| `@typescript-eslint/parser`        | `2.9.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.7.1` |
| `node`                             | `12.10.0` |
| `yarn`                              | `1.19.2` |

**Comment by bradzacher** [comment]
Hmm. This is a really really weird behaviour within the typescript typechecker.
I'm not sure if we can detect this.

Without the non-null assertion, the type of `a` is `never`, however, with the operator, typescript instead reports the type of `a` as `A`.

Better described by the on-hover information:
![image](https://user-images.githubusercontent.com/7462525/69835857-63f8cc80-11f9-11ea-8bc9-493c2846bbf1.png)

![image](https://user-images.githubusercontent.com/7462525/69835848-5cd1be80-11f9-11ea-8a4f-33642d55ecbe.png)

The reason the rule reports on the usage of the `!`, is because typescript tells us the type is `A`, which is non-nullable.

This probably deserves a bug report of some kind on the typescript side..

**Comment by Skalman** [comment]
Huh, weird. Would you like me to report the TypeScript bug?

**Comment by bradzacher** [comment]
https://github.com/microsoft/TypeScript/issues/35431 has been closed and the fix will be released with the next TS version
